### ISSUE TITLE
x,y dimension mismatch in Map3D and Extrapolator

### ISSUE THREAD

**Issue by wtbarnes** [sunpy/solarbextrapolation#7]
There is a problem in the way that Extrapolator and Map3D translate the (nx,ny) shape of the magnetogram boundary to the (nx,ny,nz) shape of 3D map that holds the extrapolated field, mainly that x and y get switched. 

For example, starting in the `Extrapolators` class in `extrapolators/base.py`, the base class of `PotentialExtrapolator`, starting on [line 95](https://github.com/sunpy/solarbextrapolation/blob/master/solarbextrapolation/extrapolators/base.py#L95), 

``` Python
self.shape = np.asarray([self.map_boundary_data.data.shape[0],
                      self.map_boundary_data.data.shape[1],
                      long(kwargs.get('zshape', 1L))])
```

Then, in `PotentialExtrapolator`, first on [line 46],(https://github.com/sunpy/solarbextrapolation/blob/master/solarbextrapolation/extrapolators/potential_field_extrapolator.py#L46),

``` Python
self.Dx = (self.xrange[1] - self.xrange[0]) / self.shape[0]
self.Dy = (self.yrange[1] - self.yrange[0]) / self.shape[1]
self.Dz = (self.zrange[1] - self.zrange[0]) / self.shape[2]
```

The problem here is that `self.map_boundary_data.shape[0]` returns the number of rows of the magnetogram data, the y-dimension, and `self.map_boundary_data.shape[1]`  returns the number of columns, the x-dimension.  (This correspondence between 0->ny and 1->nx can be checked by creating a SunPy map and then comparing the shape and "naxis1" and "naxis2" parameters.)

However, when `self.Dx` and `self.Dy` are calculated from this shape parameter, it is assumed that that `self.shape[0]` corresponds to nx and `self.shape[1]` to ny. 

Furthermore, the same problem occurs in the `Map3D` class in `map3dclasses.py` starting on [line 52](https://github.com/sunpy/solarbextrapolation/blob/master/solarbextrapolation/map3dclasses.py#L52) through [line 70](https://github.com/sunpy/solarbextrapolation/blob/master/solarbextrapolation/map3dclasses.py#L70).

**Comment by Alex-Ian-Hamilton** [comment]
OK, that's a bit of an issue, does explain some of the problems I was having in the early stages of the project though.
I'm currently just throwing the question out to people on the preferred index order of the Map3D class before doing anything.
I get back to you ASAP.

**Comment by wtbarnes** [comment]
Fair enough. It seems pretty logical to always assume a `shape` parameter uses the order (nx,ny,nz). From what I can tell, this is pretty consistently true throughout the package. The problem arises when shape is pulled straight from `np.array().shape` in which case it is (ny,nx,nz).

I have made what I _think_ are all of the necessary fixes (really not that many) on my fork. I can submit a PR and people can have a look over the changes and decide where to go from there.

**Comment by Alex-Ian-Hamilton** [comment]
Well apparently the (y, z) order is because of numpy, though I'm hazy on the details at this stage.
I intend to raise it in the issue at the SunPy meeting tomorrow and get an idea of the consensus.
Be brill to have your PR as a place we can all discuss.
Admittedly Cadair and I were tending towards the idea of moving to the (z, y, x, v) order (where v are the vectors of the field), but I'm simply not an expert on the use of 3D arrays to store data so I'm not sure the best route.

**Comment by wtbarnes** [comment]
Great. Yes I'm sure others will have thoughts on this as well. I'll try to submit the PR before the meeting tomorrow.

**Comment by wtbarnes** [comment]
Fixed by #8
